### ISSUE TITLE
Placeholder for ids is not consistent between services

### ISSUE THREAD

**Issue by metaschell** [reconciliation-api/specs#39]
The placeholders used to embed an entity id in the service URLs defined in the manifest differ between the services. These should be consistent, ideally choosing `{{id}}` over `${id}` to avoid pitfalls in clients/language interpreting `$` as replacement character within strings (which is kind of the point here as well, but still using this without explicitly providing the character, e.g. in a bash with CURL might lead to subtle and hard to spot errors...)

The preview service (section 5.1) uses `{{id}}` as placeholder in the service description url:
https://reconciliation-api.github.io/specs/latest/#preview-metadata

The flyout service (sections 6.1 and 6.5) uses `${id}` as placeholder in the service description url:
https://reconciliation-api.github.io/specs/latest/#suggest-metadata
https://reconciliation-api.github.io/specs/latest/#flyout-services

This also affects the Manifest JSON schema defined in section A.1
https://reconciliation-api.github.io/specs/latest/#manifest-schema

**Comment by wetneb** [comment]
That's also something I have found annoying. It would be great to unify this.

Also, we should make it clearer how URL escaping works in these patterns. If an id contains special characters, are they going to be escaped before being inserted in the template?

**Comment by metaschell** [comment]
Escaping should be defined, especially when returning IRIs/URIs as IDs.

We are looking to implement the Reconciliation API with linked data, so the ID is always an IRI. Or rather the local name of an IRI as I understand it, as the namespace is declared in the service manifest? Which is a bit of a problem when searching through data from mixed sources which might have mixed/different namespaces...

**Comment by wetneb** [comment]
If you are using entire IRIs as IDs, then you should be able to use templates without any prefix or suffix - that relies on the assumption that ids are not escaped when inserted in templates.

The `identifierSpace` defined in the manifest does not have any concrete influence on how ids are understood. You can just use any URI that describes your "identifier space" - it does not need to be a namespace in the sense of RDF.

On top of my head I don't remember exactly what OpenRefine currently does concerning escaping. (It's probably not worth following - we just need to think about what the specs should be in an ideal world and then update OpenRefine to follow that.)

**Comment by metaschell** [comment]
That sounds like a good approach :-)

**Comment by wetneb** [comment]
In OpenRefine we are considering introducing escaping for the ids, so that would potentially break @metaschell's use case of using full URIs as ids:
https://github.com/OpenRefine/OpenRefine/issues/3730
Can we think of a solution? Ideally we should agree on something here, update the specs, and then OpenRefine would follow.

**Comment by fsteeg** [comment]
I think the approach to treat the `{{id}}` as a URI component as implemented in https://github.com/OpenRefine/OpenRefine/pull/3731 is good. It works with the basic use case described by @metaschell in https://github.com/reconciliation-api/specs/issues/39#issuecomment-587487259:

> We are looking to implement the Reconciliation API with linked data, so the ID is always an IRI. Or rather the local name of an IRI as I understand it, as the namespace is declared in the service manifest?

This is also what we do in [our GND service](https://lobid.org/gnd/reconcile), where we use the identifierSpace `https://d-nb.info/gnd/` and IDs like `118624822` to create linked data URIs like https://d-nb.info/gnd/118624822. With the URI component encoding from https://github.com/OpenRefine/OpenRefine/pull/3731 this should also work for hash-based linked data URIs like https://d-nb.info/standards/elementset/gnd#AuthorityResource.

For the second part of https://github.com/reconciliation-api/specs/issues/39#issuecomment-587487259 (and what @wetneb described in https://github.com/OpenRefine/OpenRefine/issues/3730#issuecomment-800158508):

> Which is a bit of a problem when searching through data from mixed sources which might have mixed/different namespaces...

I think actually supporting full URIs as IDs is conceptually quite a different approach from what is specified so far, in particular regarding the `identifierSpace`. Would that be empty in a service where IDs are full URIs? The `identifierSpace` is quite central in data extension (https://reconciliation-api.github.io/specs/latest/#data-extension-responses):

>  If the property values are entities, their identifiers MUST be in the service's identifier space.

This allows the client to offer further extension on extended entities.

I also assumed that services with a common `identifierSpace` could provide interoperability. E.g. when working with the GND, I could try data extension with any service using the `https://d-nb.info/gnd/` identifierSpace (if there were multiple in the future).

Basically my understanding of the spec is that a single reconciliation service is responsible for a single `identifierSpace`. So reconciliation in different/mixed namespaces would require different reconciliation services.

I'm not saying this is the only right way, and that using full URIs as IDs is something we can't do, but I think that's probably a separate, and larger issue. For the current spec, I think using `{{id}}` consistently and defining it as something that will be URL component encoded (as in https://github.com/OpenRefine/OpenRefine/pull/3731) is good. I can look into suggesting a spec update if you agree @wetneb.

**Comment by fsteeg** [comment]
I've opened https://github.com/reconciliation-api/specs/pull/71 for the encoding part of this issue, which can be resolved without changing the template syntax. We can then focus on the actual syntax consistency issue of `{{id}}` vs. `${id}` vs. `{id}` (the latter is used in [OpenAPI](https://github.com/OAI/OpenAPI-Specification/blob/main/versions/3.1.0.md#path-templating), which was brought up by @edgardmarx on the [mailing list](https://lists.w3.org/Archives/Public/public-reconciliation/2021Apr/0008.html)), and its implications, in particular how to handle breaking changes in the spec.

**Comment by Robsteranium** [comment]
Perhaps the syntax could follow [URI Template (RFC6570)](https://datatracker.ietf.org/doc/html/rfc6570)? Of course the whole specification isn't required here but it at least provides a standard to work from. The expectations about escaping and interpolation are clear and well-documented. You can do stuff like `{+id}` if the `id` is already a URI.

**Comment by wetneb** [comment]
Hey @Robsteranium that sounds like a really really good idea!

**Comment by wetneb** [comment]
I realized last week that the update in #71 is incompatible with [the OpenCorporates endpoint](https://opencorporates.com/reconcile). They use the following view URL template: `https://opencorporates.com\{{id}}`, and the service uses identifiers such as `/companies/de/U1104_HRB22512`, which then gives you URL such as https://opencorporates.com/companies/de/U1104_HRB22512. With the escaping we have introduced, the `/` characters in the ids get escaped, which breaks the URLs.

I think we need to follow @Robsteranium's suggestion here, because I feel uncomfortable with breaking such a use case.
